Sources/RayTracer/RayTracer.Imaging/Png/PngDecoder.cs

```diff
@@ -0,0 +1,162 @@
+using System;
+using System.IO;
+
+namespace RayTracer.Imaging.Png;
+
+public static class PngDecoder
+{
+    public static void Decode(MemoryStream dataStream, PngHeader pngHeader)
+    {
+        if (pngHeader.InterlaceMethod == 1) // Adam7
+        {
+            throw new NotImplementedException("Interlaced PNGs are not supported.");
+        }
+        var bytesPerScanline = GetBytesPerScanline(pngHeader);
+
+        var data = dataStream.GetBuffer();
+
+        for (int i = 0; i < pngHeader.Height; i++)
+        {
+            var filterType = (FilterType)data[i * (bytesPerScanline + 1)];
+            ReverseFilter(filterType, data, i, bytesPerScanline);
+        }
+        
+        dataStream.Position = 0;
+    }
+
+    private static byte GetSamplesPerPixel(PngHeader header)
+    {
+        switch (header.PngColorType)
+        {
+            case PngColorType.None:
+                return 1;
+            case PngColorType.PaletteUsed:
+                return 1;
+            case PngColorType.ColorUsed:
+                return 3;
+            case PngColorType.AlphaChannelUsed:
+                return 1;
+            case PngColorType.ColorUsed | PngColorType.AlphaChannelUsed:
+                return 4;
+            default:
+                return 0;
+        }
+    }
+
+    private static byte GetBytesPerPixel(PngHeader header)
+    {
+        var bitDepthCorrected = (header.BitDepth + 7) / 8;
+        var samplesPerPixel = GetSamplesPerPixel(header);
+
+        return (byte)(bitDepthCorrected * samplesPerPixel);
+    }
+
+    private static int GetBytesPerScanline(PngHeader header)
+    {
+        var width = header.Width;
+
+        switch (header.BitDepth)
+        {
+            case 1:
+                return (int)((width + 7) / 8);
+            case 2:
+                return (int)((width + 3) / 4);
+            case 4:
+                return (int)((width + 1) / 2);
+            case 8:
+            case 16:
+                return (int)(width * GetSamplesPerPixel(header) * (header.BitDepth / 8));
+            default:
+                return 0;
+        }
+    }
+
+    private static void ApplyFilter(FilterType filterType, Span<byte> data, int rawId, int bytesPerScanline)
+    {
+        int rawStartIndex = rawId * (bytesPerScanline + 1) + 1;
+        int previousRawStartIndex = (rawId - 1) * (bytesPerScanline + 1) + 1;
+        
```

Comment:
?

Reply:
For filtering methods I need previous and current raw, so that's their indexes in the stream

Reply:
But where is rest of the method? 😅